### ISSUE TITLE
Fix S3949 FP: Do not report inside GetHashCode

### ISSUE THREAD

**Issue by Bouke** [SonarSource/sonar-dotnet#3491]
### Description

Visual Studio 2019 can generate implementations for `GetHashCode`. The generated code uses random seed values, which might overflow `int`. This is not a problem as the actual numerical value of the result is not what we're interested in.

### Repro steps

```c#
        public override int GetHashCode()
        {
            int hashCode = -1205156623;
            hashCode = hashCode * -1521134295 + TotalAllocatedMemorySize.GetHashCode();
            hashCode = hashCode * -1521134295 + SurvivedMemorySize.GetHashCode();
            return hashCode;
        }
```

### Expected behavior

No warning raised.

### Actual behavior

S3949 is raised as the result overflows.

### Known workarounds

-

### Related information

* Visual Studio Version: 2019 (16.6.5)
* https://community.sonarsource.com/t/s3949-for-gethashcode/29414

### ToDo:

* [x] C#
* [x] VB

**Comment by pavel-mikula-sonarsource** [comment]
Hi @Bouke,

Any hash algorithm is generally fine with overfloating integer numbers. And the purpose of this rule is to detect such situations. Marking this as False Positive in your SonarQube/SonarCloud would be the most appropriate solution in this case.

cc @loic-joly-sonarsource in case the CBDE has build-in whitelisting mechanism for known hash functions.

**Comment by Bouke** [comment]
It would still report a "major bug" whenever we do pull request annotations, which is highly annoying, confusing reviewers and just teaches people to ignore sonar warnings. I suggest to reopen this issue and work on whitelisting overflows in hash functions.

**Comment by Bouke** [comment]
Additionally; suppressing the message in-code doesn't work:

```
[SuppressMessage("Major Code Smell", "S3949:Calculations should not overflow", Justification = "Designed to overflow")]
        public override int GetHashCode() { ... }
```

**Comment by pavel-mikula-sonarsource** [comment]
You can disable the rule for that project if it doesn't fit your needs.
There's also IDE0070 suggesting to use `return System.HashCode.Combine(TotalAllocatedMemorySize, SurvivedMemorySize);` instead. You can configure this one to fail your Quality Gate.

We have no control of `System.Diagnostics.CodeAnalysis.SuppressMessage` attribute. You can contact Microsoft with this issue.

**Comment by Bouke** [comment]
[System.HashCode doesn't exist in .Net Framework](https://docs.microsoft.com/en-us/dotnet/api/system.hashcode?view=dotnet-plat-ext-3.1). So IDE0070 doesn't apply in our case.

> We have no control of System.Diagnostics.CodeAnalysis.SuppressMessage attribute. You can contact Microsoft with this issue.

Well this specific analysis apparently happens in a separate process, probably outside of Razor Analyzer's reach. I suspect that causes `SuppressMessage` to have no effect. So that means it isn't Microsoft's issue, but yours, as this out-of-process analyzer doesn't respect `SuppressMessage`.

**Comment by andrei-epure-sonarsource** [comment]
The `SuppressMessage` issue should be a different issue than this @Bouke
